File: backend/app/schemas/resources/resource.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, HttpUrl, Field


class ResourceCreateFromUrl(BaseModel):
    url: HttpUrl
    category_id: int
    is_public: Optional[bool] = False
    manual_weight: Optional[int] = None
    model_config = ConfigDict(from_attributes=True)


class ResourceUpdateRequest(BaseModel):
    title: Optional[str] = None
    summary: Optional[str] = None
    platform: Optional[str] = None
    thumbnail: Optional[str] = None
    category_id: Optional[int] = None
    difficulty: Optional[int] = None
    tags: Optional[dict[str, Any]] = None
    raw_meta: Optional[dict[str, Any]] = None
    manual_weight: Optional[int] = None
    model_config = ConfigDict(from_attributes=True)


class VideoInfo(BaseModel):
    duration: Optional[int] = None
    channel: Optional[str] = None
    video_id: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)


class DocInfo(BaseModel):
    doc_type: Optional[str] = None
    version: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)


class ArticleInfo(BaseModel):
    publisher: Optional[str] = None
    published_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)


class ResourceResponse(BaseModel):
    id: int
    resource_type: str
    platform: Optional[str] = None
    title: str
    summary: Optional[str] = None
    source_url: str
    thumbnail: Optional[str] = None
    category_id: int
    category_name: Optional[str] = None
    difficulty: Optional[int] = None
    tags: Optional[dict[str, Any]] = None
    raw_meta: Optional[dict[str, Any]] = None
    is_system_public: Optional[bool] = None

    manual_weight: Optional[int] = None
    behavior_weight: Optional[int] = None
    effective_weight: Optional[int] = None
    added_at: Optional[datetime] = None
    last_opened: Optional[datetime] = None
    open_count: Optional[int] = None
    completion_status: Optional[bool] = None

    community_score: Optional[int] = None
    save_count: Optional[int] = None
    trending_score: Optional[int] = None

    created_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)


class ResourceDetailResponse(ResourceResponse):
    video: Optional[VideoInfo] = None
    doc: Optional[DocInfo] = None
    article: Optional[ArticleInfo] = None


class ResourceAttachResponse(BaseModel):
    already_exists: bool
    resource: ResourceResponse

    model_config = ConfigDict(from_attributes=True)

File: app/src/api/resource.ts
import request from '../utils/request'

export interface UrlExtractResponse {
  title: string
  description?: string | null
  thumbnail_url?: string | null
  author?: string | null
  publish_date?: string | null
  video_id?: string | null
  duration_seconds?: number | null
  platform?: string | null
  chapters?: ChapterItem[]
}

export function extractVideoMetadata(url: string) {
  return request.post<UrlExtractResponse, UrlExtractResponse>('/resources/extract', { url })
}

export interface DbResource {
  id: number
  resource_type: string
  platform: string
  title: string
  summary: string | null
  source_url: string
  thumbnail: string | null
  category_id: number | null
  difficulty: string | null
  tags: object
  created_at: string
  updated_at: string
  category_name?: string
  is_system_public?: boolean

  manual_weight?: number | null
  behavior_weight?: number | null
  effective_weight?: number | null
  added_at?: string | null
  last_opened?: string | null
  open_count?: number | null
  completion_status?: boolean | null

  community_score?: number | null
  save_count?: number | null
  trending_score?: number | null
  user_seq?: number | null
}

// 添加资源的请求参数
export interface CreateResourceRequest {
  url: string
  category_id: number
  is_system_public: boolean
}

// 添加资源的响应
export interface CreateResourceResponse {
  resource: DbResource
}

// 添加资源


export interface ChapterItem {
  start_seconds: number
  timestamp: string
  title: string
  description?: string | null
}

export interface DbResourceDetail extends DbResource {
  video?: {
    duration?: number | null
    channel?: string | null
    video_id?: string | null
  } | null
  doc?: {
    doc_type?: string | null
    version?: string | null
  } | null
  article?: {
    publisher?: string | null
    published_at?: string | null
  } | null
}

export function listMyResources() {
  return request.get<DbResource[], DbResource[]>('/resources/me')
}

export function listResources() {
  return request.get<DbResource[], DbResource[]>('/resources')
}

export function createMyResourceFromUrl(
  url: string,
  payload?: { category_id: number; is_public?: boolean; manual_weight?: number },
) {
  return request.post<DbResource, DbResource>('/resources/me', { url, ...(payload || {}) })
}

export function addPublicResourceToMyResources(resourceId: number) {
  return request.post<DbResource, DbResource>(`/resources/me/${resourceId}`, {})
}

export type AddToMyResourcesResult = {
  already_exists: boolean
  resource: DbResource
}

export function addPublicResourceToMyResourcesWithStatus(resourceId: number) {
  return request.post<AddToMyResourcesResult, AddToMyResourcesResult>(`/resources/me/${resourceId}/attach`, {})
}

export function addPublicResourceToMyResourcesWithStatusAndWeight(resourceId: number, payload?: { manual_weight?: number }) {
  return request.post<AddToMyResourcesResult, AddToMyResourcesResult>(`/resources/me/${resourceId}/attach`, payload || {})
}

export function deleteMyResource(resourceId: number) {
  return request.delete(`/resources/me/${resourceId}`)
}

export function deleteResource(resourceId: number) {
  return request.delete(`/resources/${resourceId}`)
}

export function updateMyResource(
  resourceId: number,
  payload: {
    title?: string
    summary?: string | null
    platform?: string | null
    thumbnail?: string | null
    difficulty?: number | null
    tags?: Record<string, any> | null
    raw_meta?: Record<string, any> | null
    manual_weight?: number | null
  },
) {
  return request.patch<DbResource, DbResource>(`/resources/me/${resourceId}`, payload)
}

export function getMyResourceDetail(resourceId: number) {
  return request.get<DbResourceDetail, DbResourceDetail>(`/resources/me/${resourceId}`)
}

export function getResourceDetail(resourceId: number) {
  return request.get<DbResourceDetail, DbResourceDetail>(`/resources/${resourceId}`)
}